File: apps/android/app/src/test/java/dev/crossbridge/android/TrustedDeviceStoreTest.kt
package dev.crossbridge.android

import dev.crossbridge.android.data.InMemoryKeyValueStore
import dev.crossbridge.android.data.TrustedDeviceStore
import dev.crossbridge.android.protocol.TrustedDevice
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class TrustedDeviceStoreTest {
    @Test
    fun saveAndLoadDevice() {
        val store = TrustedDeviceStore(InMemoryKeyValueStore())
        val device = trustedDevice(deviceId = "pc_1")

        store.saveTrustedDevice(device)

        assertEquals(listOf(device), store.loadTrustedDevices())
    }

    @Test
    fun deduplicatesByDeviceId() {
        val store = TrustedDeviceStore(InMemoryKeyValueStore())
        store.saveTrustedDevice(trustedDevice(deviceId = "pc_1", deviceName = "Old PC"))
        store.saveTrustedDevice(
            trustedDevice(
                deviceId = "pc_1",
                deviceName = "Updated PC",
                pairedAt = 200,
                lastSeenAt = 300
            )
        )

        val devices = store.loadTrustedDevices()

        assertEquals(1, devices.size)
        assertEquals("Updated PC", devices.single().deviceName)
        assertEquals(300L, devices.single().lastSeenAt)
    }

    @Test
    fun updatingExistingDeviceChangesLastSeenAt() {
        var nowMs = 500L
        val store = TrustedDeviceStore(InMemoryKeyValueStore()) { nowMs }
        store.saveTrustedDevice(trustedDevice(deviceId = "pc_1", lastSeenAt = null))

        nowMs = 900L
        store.saveTrustedDevice(trustedDevice(deviceId = "pc_1", lastSeenAt = null))

        assertEquals(900L, store.loadTrustedDevices().single().lastSeenAt)
    }

    @Test
    fun removeDevice() {
        val store = TrustedDeviceStore(InMemoryKeyValueStore())
        store.saveTrustedDevice(trustedDevice(deviceId = "pc_1"))
        store.saveTrustedDevice(trustedDevice(deviceId = "pc_2", pairedAt = 200))

        store.removeTrustedDevice("pc_1")

        val devices = store.loadTrustedDevices()
        assertEquals(1, devices.size)
        assertEquals("pc_2", devices.single().deviceId)
    }

    @Test
    fun malformedStoredJsonIsIgnoredSafely() {
        val keyValueStore = InMemoryKeyValueStore()
        keyValueStore.putString("trustedDevices", "{not valid json")
        val store = TrustedDeviceStore(keyValueStore)

        assertTrue(store.loadTrustedDevices().isEmpty())
    }

    private fun trustedDevice(
        deviceId: String,
        deviceName: String = "Windows PC",
        pairedAt: Long = 100,
        lastSeenAt: Long? = null
    ): TrustedDevice {
        return TrustedDevice(
            deviceId = deviceId,
            deviceName = deviceName,
            platform = "windows",
            publicKey = "pc_public_key",
            pairedAt = pairedAt,
            lastSeenAt = lastSeenAt
        )
    }
}

File: apps/windows/src/services/trustedDeviceStore.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";
import { createTrustedDeviceStore } from "./trustedDeviceStore.js";
import type { TrustedDevice } from "@crossbridge/protocol";

class MemoryStorage {
  private readonly values = new Map<string, string>();

  getItem(key: string): string | null {
    return this.values.get(key) ?? null;
  }

  setItem(key: string, value: string): void {
    this.values.set(key, value);
  }

  removeItem(key: string): void {
    this.values.delete(key);
  }

  setRaw(key: string, value: string): void {
    this.values.set(key, value);
  }
}

const storageKey = "crossbridge.trustedDevices.v1";

const pixel: TrustedDevice = {
  deviceId: "android_pixel",
  deviceName: "Pixel",
  platform: "android",
  publicKey: "YW5kcm9pZC1wdWJsaWMta2V5",
  pairedAt: 1_000
};

afterEach(() => {
  vi.restoreAllMocks();
});

describe("trustedDeviceStore", () => {
  it("saves, loads, removes, and clears devices through the async browser fallback", async () => {
    const storage = new MemoryStorage();
    const store = createTrustedDeviceStore(storage);

    await store.saveTrustedDevice(pixel);
    expect(await store.loadTrustedDevices()).toEqual([pixel]);

    await store.removeTrustedDevice(pixel.deviceId);
    expect(await store.loadTrustedDevices()).toEqual([]);

    await store.saveTrustedDevice(pixel);
    await store.clearTrustedDevices();
    expect(await store.loadTrustedDevices()).toEqual([]);
  });

  it("deduplicates devices by deviceId and refreshes lastSeenAt", async () => {
    const storage = new MemoryStorage();
    const store = createTrustedDeviceStore(storage);

    await store.saveTrustedDevice(pixel);
    vi.spyOn(Date, "now").mockReturnValue(5_000);
    await store.saveTrustedDevice({
      ...pixel,
      deviceName: "Pixel 8"
    });

    expect(await store.loadTrustedDevices()).toEqual([
      {
        ...pixel,
        deviceName: "Pixel 8",
        lastSeenAt: 5_000
      }
    ]);
  });

  it("ignores malformed stored entries safely", async () => {
    const storage = new MemoryStorage();
    const store = createTrustedDeviceStore(storage);

    storage.setRaw(storageKey, JSON.stringify([
      pixel,
      { ...pixel, deviceId: "" },
      { deviceId: "android_bad" },
      "not-a-device"
    ]));

    expect(await store.loadTrustedDevices()).toEqual([pixel]);

    storage.setRaw(storageKey, "{not json");
    expect(await store.loadTrustedDevices()).toEqual([]);
  });

  it("does not crash when storage is unavailable", async () => {
    const store = createTrustedDeviceStore(undefined);

    await expect(store.saveTrustedDevice(pixel)).resolves.toBeUndefined();
    expect(await store.loadTrustedDevices()).toEqual([]);
    await expect(store.removeTrustedDevice(pixel.deviceId)).resolves.toBeUndefined();
    await expect(store.clearTrustedDevices()).resolves.toBeUndefined();
  });
});
